src/config.cpp

```diff
@@ -16,120 +16,231 @@ ONNXTensorElementDataType TranslateTensorType(std::string_view value) {
+    } else if (name == "head_size") {
+      v_.head_size = static_cast<int>(value);
+    } else
+      throw JSON::unknown_value_error{};
+  }
+
+  Element& OnObject(std::string_view name) override {
+    if (name == "inputs") {
+      return inputs_;
+    }
+    if (name == "outputs") {
+      return outputs_;
+    }
+    throw JSON::unknown_value_error{};
+  }
+
+ private:
+  Config::Model::Decoder& v_;
+  Inputs_Element inputs_{v_.inputs};
+  Outputs_Element outputs_{v_.outputs};
+};
+
+struct Model_Element : JSON::Element {
+  explicit Model_Element(Config::Model& v) : v_{v} {}
+
+  void OnString(std::string_view name, std::string_view value) override {
+    if (name == "type") {
+      v_.type = value;
     } else if (name == "logits_type") {
-      model_.logits_type = TranslateTensorType(value);
+      v_.logits_type = TranslateTensorType(value);
     } else if (name == "kv_type") {
-      model_.kv_type = TranslateTensorType(value);
-    } else {
-      throw std::runtime_error("Unknown name: " + std::string(name));
-    }
+      v_.kv_type = TranslateTensorType(value);
+    } else
+      throw JSON::unknown_value_error{};
   }
 
   void OnNumber(std::string_view name, double value) override {
     if (name == "vocab_size") {
-      model_.vocab_size = static_cast<int>(value);
-    } else if (name == "hidden_size" || name == "n_embed") {
-      model_.hidden_size = static_cast<int>(value);
-    } else if (name == "num_attention_heads" || name == "num_heads" || name == "n_head") {
-      model_.num_attention_heads = static_cast<int>(value);
-    } else if (name == "num_hidden_layers" || name == "num_layers" || name == "n_layer") {
-      model_.num_hidden_layers = static_cast<int>(value);
+      v_.vocab_size = static_cast<int>(value);
+    } else if (name == "context_length") {
+      v_.context_length = static_cast<int>(value);
+    } else if (name == "pad_token_id") {
+      v_.pad_token_id = static_cast<int>(value);
+    } else if (name == "eos_token_id") {
+      v_.eos_token_id = static_cast<int>(value);
+    } else if (name == "bos_token_id") {
+      v_.bos_token_id = static_cast<int>(value);
+    } else if (name == "decoder_start_token_id") {
+      v_.decoder_start_token_id = static_cast<int>(value);
+    } else if (name == "sep_token_id") {
+      v_.sep_token_id = static_cast<int>(value);
+    } else
+      throw JSON::unknown_value_error{};
+  }
+
+  Element& OnObject(std::string_view name) override {
+    if (name == "encoder_decoder_init") {
+      return encoder_decoder_init_;
     }
+    if (name == "decoder") {
+      return decoder_;
+    }
+    throw JSON::unknown_value_error{};
   }
 
  private:
-  Config::Model& model_;
+  Config::Model& v_;
+  EncoderDecoderInit_Element encoder_decoder_init_{v_.encoder_decoder_init};
+  Decoder_Element decoder_{v_.decoder};
 };
 
-struct Root_Element : JSON::Element {
-  explicit Root_Element(Config& config) : config_{config} {}
+struct Search_Element : JSON::Element {
```

Comment:
nit: name it as Generato_Element for consistency

Reply:
Hmm, the values here are for how to do the search (beamsearch/greedysearch).
But I can see how they're also how the generation loop works.
But the GeneratorParams also has the model inputs.
Up to you, I can call this Generator_Element, but it's not directly related to GeneratorParams.